python/telemetry/snet.py

```diff
@@ -168,10 +102,105 @@ def _decode(self, obj, context, path=None):
+    'Hailing' / Flag,
+    'UdFl1' / Flag,
+    'PduLength' / BitsInteger(10),
+    'CRC13' / BitsInteger(13),
+    'CRC5' / BitsInteger(5),
+    Padding(2)
+    )
 
-snet = Struct(
-    'header' / SNETFrameHeader,
-    'telemetry' / Switch(lambda c: (c.header.FCIDMajor, c.header.FCIDSub), {
+# Mailbox in SNET satellites
+# There are three main fields defining the Mailbox inside SNET satellites:
+#
+# Name        ByteLength  ParamType   TargetType    Interpretation
+# SysTime     4           uint32_t    float         scale: 0,001 unit: s
+# MemDataType 1           uint8_t     uint8_t       values: 0, 1, 2, 3, 4, 5
+# MemData32   32          uint8_t     Array         format: Array
+#
+# Notes:
+# * `MemDataType` refers to the slot of memory in the satellite where
+# the information is stored, which slot of memory.
+# * `MemData32` is an Array of bytes, where each byte corresponds to an ascii
+# character code. The transmission of words is done by Big-Endian order.
+
+MailboxReceiveTelemetry = Struct(
+    'SysTime' / BitsInteger(32),
+    'MemDataType' / BitsInteger(8),
+    'MemData32' / BitsInteger(256)
+    )
+
+MailboxSendTelemetry = Struct(
+    'MemDataType' / BitsInteger(8),
+    'MemData32' / BitsInteger(256)
+    )
+
+
+class TimeAdapter(Adapter):
+    def _encode(self, obj, context, path=None):
+        d = int((obj - datetime.datetime(2000, 1, 1))*2)
+        return Container(
+            days=d.days,
+            milliseconds=d.seconds * 1000 + d.microseconds / 1000)
+
+    def _decode(self, obj, context, path=None):
+        return (datetime.datetime(2000, 1, 1)
+                + datetime.timedelta(seconds=float(obj)/2.0))
+
+
+TimeStamp = TimeAdapter(BitsInteger(32, swapped=True))
+
+PDUFrameHeaderExtension = BitStruct(
+    'VersNo' / BitsInteger(2),
+    'DFCID' / BitsInteger(2),
+    'ExtensionRFU' / BitsInteger(4),
+    'ChannelInfo' / BitsInteger(8),
+    'QoS' / Flag,
+    'PDUTypeID' / Flag,
+    'ARQ' / Flag,
+    'ControlRFU' / BitsInteger(5),
+    'TimeTagSub' / BitsInteger(16),
+    'SCID' / BitsInteger(10),
+    'SeqNo' / BitsInteger(14)
+    )
+
+PDUFrameHeader = BitStruct(
+    Const(0b111100110101000000, BitsInteger(18)),
+    'CRC' / BitsInteger(14),
+    'FCIDMajor' / BitsInteger(6),
+    'FCIDSub' / BitsInteger(10),
+    'Urgent' / Flag,
+    'Extended' / Flag,
+    'CheckCRC' / Flag,
+    'Multiframe' / Flag,
+    'TimeTaggedSetting' / Flag,
+    'TimeTagged' / Flag,
+    'DataLength' / BitsInteger(10),
+    'TimeTag' / If(lambda c: c.TimeTagged, TimeStamp),
+    'Extension' / If(lambda c: c.Extended, PDUFrameHeaderExtension),
+    'UserData' / Switch(lambda c: (c.FCIDMajor, c.FCIDSub), {
         (0, 0): ADCSTelemetry,
         (9, 0): EPSTelemetry,
-    }, default=Pass))
+        (54, 17): MailboxReceiveTelemetry,  # COM0
+        (54, 117): MailboxReceiveTelemetry,  # COM1
+        (56, 34): MailboxSendTelemetry,  # COM0
+        (56, 134): MailboxSendTelemetry,  # COM1
+    },
+      default=Pass))
```

Comment:
The user data should not be part of the header, right?

Reply:
yes, it is separated now